tools/releaser/main.go

```diff
@@ -0,0 +1,221 @@
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func log(msg string, format ...any) {
+	fmt.Fprintf(flag.CommandLine.Output(), msg+"\n", format...)
+}
+
+func run(ctx context.Context, stderr *os.File) error {
+	var (
+		goVersion    string
+		repoRoot     string
+		skipUpgrades bool
+		tag          string
+	)
+
+	flag.StringVar(&goVersion, "go_version", "", "go version for go.mod")
+	flag.StringVar(&repoRoot, "repo_root", os.Getenv("BUILD_WORKSPACE_DIRECTORY"), "root directory of bazel-zig-cc repo")
+	flag.StringVar(&tag, "tag", "", "tag for this release")
+	flag.BoolVar(&skipUpgrades, "skip_upgrades", false, "skip upgrade checks (testing only)")
+
+	flag.Usage = func() {
+		fmt.Fprint(flag.CommandLine.Output(), `usage: bazel run //tools/releaser -- -go_version <version> -tag <tag>
+
+This utility is intended to handle many of the steps to release a new version.
+
+`)
+		flag.PrintDefaults()
+	}
+
+	flag.Parse()
+
+	if tag == "" {
+		return fmt.Errorf("ERROR: tag is required")
+	}
+
+	if !tagRegexp.MatchString(tag) {
+		return tagErr
+	}
+
+	var goVersionArgs []string
+	if goVersion != "" {
+		versionParts := strings.Split(goVersion, ".")
+		if len(versionParts) < 2 {
+			flag.Usage()
+			return errors.New("please provide a valid Go version")
+		}
+		if minorVersion, err := strconv.Atoi(versionParts[1]); err != nil {
+			return fmt.Errorf("%q is not a valid Go version", goVersion)
+		} else if minorVersion > 0 {
+			versionParts[1] = strconv.Itoa(minorVersion - 1)
+		}
+		goVersionArgs = append(goVersionArgs, "-go", goVersion, "-compat", strings.Join(versionParts, "."))
+	}
+
+	workspacePath := path.Join(repoRoot, "WORKSPACE")
+
+	log("Running pre-release checks:")
+	if skipUpgrades {
+		log("- SKIPPING: go update commands")
+	} else {
+		log("- initial go update commands")
+		initialCommands := []struct {
+			cmd  string
+			args []string
+		}{
+			{cmd: "go", args: []string{"get", "-t", "-u", "./..."}},
+			{cmd: "tools/mod-tidy", args: goVersionArgs},
+		}
+		for _, c := range initialCommands {
+			cmd := exec.CommandContext(ctx, c.cmd, c.args...)
+			cmd.Dir = repoRoot
+			if out, err := cmd.CombinedOutput(); err != nil {
+				return fmt.Errorf(
+					"ERROR: running %s:%w\n%s",
+					strings.Join(append([]string{c.cmd}, c.args...), " "),
+					err,
+					out,
+				)
+			}
+		}
+	}
+
+	workspace, err := os.OpenFile(workspacePath, os.O_RDWR, 0o644)
+	if err != nil {
+		return err
+	}
+	defer workspace.Close()
+
+	log("- gazelle")
+	cmd := exec.CommandContext(ctx, "bazel", "run", "//:gazelle")
+	cmd.Dir = repoRoot
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("ERROR: gazelle failed: %w\n%s", err, out)
+	}
+
+	log("- checking if repository is clean")
+	cmd = exec.Command("git", "diff", "--stat", "--exit-code")
+	cmd.Dir = repoRoot
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf(
+			"ERROR: repository is dirty. Please commit your changes to continue:\n%s",
+			out,
+		)
+	}
```

Comment:
Are these two steps necessary during a release? They are already checked at CI for each PR, right?

Reply:
This is part of the (updated) dependency resolution, which is not done during every CI run.

Reply:
Dependency resolution is through `go mod tidy && gazelle update-repos`. Do you see a case when `bazel run //:gazelle` would modify the build files after those two commands?